frost-core/src/keys/refresh.rs

```diff
@@ -0,0 +1,119 @@
+//! Refresh Shares
+//!
+//! Implements the functionality to refresh a share. This requires the participation
+//! of all the remaining signers. This can be done using a Trusted Dealer or
+//! DKG (not yet implemented)
+
+use std::collections::BTreeMap;
+
+use crate::{
+    keys::{
+        generate_coefficients, generate_secret_shares, validate_num_of_signers,
+        CoefficientCommitment, PublicKeyPackage, SigningKey, SigningShare, VerifyingShare,
+    },
+    Ciphersuite, CryptoRng, Error, Field, Group, Identifier, RngCore, Scalar,
+};
+
+use super::{SecretShare, VerifiableSecretSharingCommitment};
+
+/// Refreshes shares using a trusted dealer
+pub fn refresh_shares_with_dealer<C: Ciphersuite, R: RngCore + CryptoRng>(
+    old_shares: BTreeMap<Identifier<C>, SecretShare<C>>,
+    old_pub_key_package: PublicKeyPackage<C>,
+    max_signers: u16,
+    min_signers: u16,
+    identifiers: &[Identifier<C>],
+    rng: &mut R,
+) -> Result<(BTreeMap<Identifier<C>, SecretShare<C>>, PublicKeyPackage<C>), Error<C>> {
```

Comment:
I think we'll need to refactor the API. Consider the regular trusted dealer: the dealer generates a bunch of `SecretShares` that are sent to the participants which convert them to `KeyPackage`s.

If they want to refresh their shares, with this API, they would need to:

- Sent their KeyPackages back to the Coordinator, which would then need to convert them back to SecretShares
- Call `refresh_shares_with_dealer()`
- Send the refreshed `SecretShares` to the participants

An easier flow would be:

- The Coordinator calls something like `generate_refresh_shares_with_dealer()` (name ideas welcome) which is basically just `split()` with a zero key
- The Coordinator sends the zero-key `SecretShares` to participants
- Each participant would them call a new e.g. `refresh_key_package()` with their current `KeyPackage` and the zero-key `SecretPackage`, add them with the logic you already wrote and return the refreshed `KeyPackage`

Another upside is that `refresh_key_package()` would also be used by the DKG refresh share.